### PR TITLE
[pytorch] [trcomp] Add support for PT 1.13 SM Training Compiler

### PR DESCRIPTION
**Pull request by BruceChang9688**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [x] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [x] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [x] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [x] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/sagemaker_tests/pytorch/training/integration/sagemaker/test_distributed_operations.py [hunk 1]
```diff
@@ -116,6 +116,8 @@ def test_dist_operations_fastai_gpu(framework_version, ecr_image, sagemaker_regi
     _, image_framework_version = get_framework_and_version_from_tag(ecr_image)
     if Version("1.9") <= Version(image_framework_version) < Version("1.13"):
         pytest.skip("Fast ai is not supported on PyTorch v1.9.x, v1.10.x, v1.11.x, v1.12.x")
+    if ecr_image.split("/")[1].startswith("pr-pytorch-trcomp-training"):
```
Comment (ShiboXing) [inline test/sagemaker_tests/pytorch/training/integration/sagemaker/test_distributed_operations.py:119]: can you make it ```if _ == "pytorch_trcomp":``` in the next commit?
sorry for the previous code suggestion

File: test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py [hunk 2]
```diff
@@ -137,7 +137,7 @@ def test_pytorch_train_dgl_cpu(pytorch_training, ec2_connection, cpu_only, py3_o
 @pytest.mark.parametrize("ec2_instance_type", PT_EC2_GPU_INSTANCE_TYPE, indirect=True)
 def test_pytorch_with_horovod(pytorch_training, ec2_connection, gpu_only, ec2_instance_type):
     _, image_framework_version = get_framework_and_version_from_tag(pytorch_training)
-    if 'trcomp' in pytorch_training and Version(image_framework_version) in SpecifierSet("<1.13.*"):
+    if 'trcomp' in pytorch_training and Version(image_framework_version) in SpecifierSet("<1.14.*"):
```
Comment (ShiboXing) [inline test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py:140]: let's use SpecifierSet("<2.0") as per the next pytorch version

File: test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py [hunk 3]
```diff
@@ -264,7 +264,7 @@ def test_pytorch_amp(pytorch_training, ec2_connection, gpu_only, ec2_instance_ty
 def test_pytorch_s3_plugin_gpu(pytorch_training, ec2_connection, gpu_only, ec2_instance_type, outside_versions_skip):
     outside_versions_skip(pytorch_training, "1.8.0", "1.12.1")
     _, image_framework_version = get_framework_and_version_from_tag(pytorch_training)
-    if 'trcomp' in pytorch_training and Version(image_framework_version) in SpecifierSet("<1.13.*"):
+    if 'trcomp' in pytorch_training and Version(image_framework_version) in SpecifierSet("<1.14.*"):
```
Comment (ShiboXing) [inline test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py:267]: let's use SpecifierSet("<2.0") as per the next pytorch version

File: test/dlc_tests/ec2/test_smdebug.py [hunk 4]
```diff
@@ -30,7 +30,7 @@ def test_smdebug_gpu(training, ec2_connection, region, ec2_instance_type, gpu_on
         pytest.skip(f"Image {training} is incompatible with instance type {ec2_instance_type}")
 
     _, image_framework_version = get_framework_and_version_from_tag(training)
-    if 'trcomp' in training and 'pytorch' in training and Version(image_framework_version) in SpecifierSet("<1.13.*"):
+    if 'trcomp' in training and 'pytorch' in training and Version(image_framework_version) in SpecifierSet("<1.14.*"):
```
Comment (ShiboXing) [inline test/dlc_tests/ec2/test_smdebug.py:33]: let's use SpecifierSet("<2.0") as per the next pytorch version

File: test/dlc_tests/ec2/test_smdebug.py [hunk 5]
```diff
@@ -64,7 +64,7 @@ def test_smprofiler_gpu(
     if test_utils.is_image_incompatible_with_instance_type(training, ec2_instance_type):
         pytest.skip(f"Image {training} is incompatible with instance type {ec2_instance_type}")
     _, image_framework_version = get_framework_and_version_from_tag(training)
-    if 'trcomp' in training and 'pytorch' in training and Version(image_framework_version) in SpecifierSet("<1.13.*"):
+    if 'trcomp' in training and 'pytorch' in training and Version(image_framework_version) in SpecifierSet("<1.14.*"):
```
Comment (ShiboXing) [inline test/dlc_tests/ec2/test_smdebug.py:67]: let's use SpecifierSet("<2.0") as per the next pytorch version

**Review by ShiboXing — CHANGES_REQUESTED** [review]

File: test/dlc_tests/sanity/test_pre_release.py [hunk 6]
```diff
@@ -791,3 +791,4 @@ def test_mxnet_training_sm_env_variables(mxnet_training):
         env_vars_to_test=env_vars,
         container_name_prefix=container_name_prefix
     )
+
```
Comment (ShiboXing) [inline test/dlc_tests/sanity/test_pre_release.py:794]: can you add back the new line here

File: test/sagemaker_tests/pytorch/training/integration/sagemaker/test_smdataparallel.py [hunk 7]
```diff
@@ -95,6 +95,8 @@ def test_smdataparallel_mnist_script_mode_multigpu(ecr_image, sagemaker_regions,
     """
     Tests SM Distributed DataParallel single-node via script mode
     """
+    if ecr_image.split("/")[1].startswith("pr-pytorch-trcomp-training"):
```
Comment (tejaschumbalkar) [inline test/sagemaker_tests/pytorch/training/integration/sagemaker/test_smdataparallel.py:98]: the ecr repo name changes in the rc build. use `endswith("pytorch-trcomp-training")`

File: test/sagemaker_tests/pytorch/training/integration/sagemaker/test_smdataparallel.py [hunk 8]
```diff
@@ -127,6 +129,8 @@ def test_smdataparallel_mnist(ecr_image, sagemaker_regions, efa_instance_type, t
     """
     Tests smddprun command via Estimator API distribution parameter
     """
+    if ecr_image.split("/")[1].startswith("pr-pytorch-trcomp-training"):
```
Comment (tejaschumbalkar) [inline test/sagemaker_tests/pytorch/training/integration/sagemaker/test_smdataparallel.py:132]: the ecr repo name changes in the rc build. use `endswith("pytorch-trcomp-training")`

File: test/sagemaker_tests/pytorch/training/integration/sagemaker/test_smdataparallel.py (hunk 7 above)
Comment (tejaschumbalkar) [reply test/sagemaker_tests/pytorch/training/integration/sagemaker/test_smdataparallel.py:98]: Also add this logic in `validate_or_skip_smdataparallel` function

File: test/sagemaker_tests/pytorch/training/integration/sagemaker/test_smdataparallel.py (hunk 8 above)
Comment (tejaschumbalkar) [reply test/sagemaker_tests/pytorch/training/integration/sagemaker/test_smdataparallel.py:132]: Also add this logic in `validate_or_skip_smdataparallel` function

File: test/sagemaker_tests/pytorch/training/integration/sagemaker/test_smdataparallel.py (hunk 7 above)
Comment (tejaschumbalkar) [inline test/sagemaker_tests/pytorch/training/integration/sagemaker/test_smdataparallel.py:98]: Use `@pytest.mark.skip_trcomp_containers` marker to skip test

File: test/sagemaker_tests/pytorch/training/integration/sagemaker/test_smdataparallel.py (hunk 8 above)
Comment (tejaschumbalkar) [inline test/sagemaker_tests/pytorch/training/integration/sagemaker/test_smdataparallel.py:132]: Use `@pytest.mark.skip_trcomp_containers` marker to skip test

File: test/sagemaker_tests/pytorch/training/integration/sagemaker/test_distributed_operations.py [hunk 9]
```diff
@@ -116,6 +116,8 @@ def test_dist_operations_fastai_gpu(framework_version, ecr_image, sagemaker_regi
     _, image_framework_version = get_framework_and_version_from_tag(ecr_image)
     if Version("1.9") <= Version(image_framework_version) < Version("1.13"):
         pytest.skip("Fast ai is not supported on PyTorch v1.9.x, v1.10.x, v1.11.x, v1.12.x")
+    if _ == "pytorch_trcomp":
```
Comment (tejaschumbalkar) [inline test/sagemaker_tests/pytorch/training/integration/sagemaker/test_distributed_operations.py:119]: Use `@pytest.mark.skip_trcomp_containers` marker to skip test

File: test/sagemaker_tests/pytorch/training/integration/sagemaker/test_smdataparallel.py [hunk 10]
```diff
@@ -31,6 +31,8 @@
 def validate_or_skip_smdataparallel(ecr_image):
     if not can_run_smdataparallel(ecr_image):
         pytest.skip("Data Parallelism is supported on CUDA 11 on PyTorch v1.6 and above")
+    if ecr_image.split("/")[1].endswith("pytorch-trcomp-training"):
```
[the remark this answers is not in the document]
Comment (BruceChang9688) [reply test/sagemaker_tests/pytorch/training/integration/sagemaker/test_smdataparallel.py:34]: I can delete this part if needed.

**Comment by ShiboXing** [comment]
standard: 529b700
rc: 626cd89
efa: eec45be

**Review by tejaschumbalkar — COMMENTED** [review]
LGTM. reverted the toml changes for merge

File: pytorch/training/docker/1.13/py3/cu117/Dockerfile.trcomp.gpu [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+########
+###   :::Update NCCL:::
+###   This is a temporary workaround to
+###   upgrade NCCL
+########
+
+# Remove Older NCCL
+RUN rm -rvf `find /usr/local | grep nccl` /usr/local/obj
+
+ENV NCCL_VERSION=2.14.3
```
Comment (Lokiiiiii) [inline pytorch/training/docker/1.13/py3/cu117/Dockerfile.trcomp.gpu:13]: Looks like the base container has the same NCCL version. Why is this reinstall necessary ?

**Comment by Lokiiiiii** [comment]
If the installation of SMDDP is broken, we should uninstall it from the container in addition to skipping the tests for the same.

**Review by Lokiiiiii — CHANGES_REQUESTED** [review]

File: pytorch/training/docker/1.13/py3/cu117/Dockerfile.trcomp.gpu (hunk 11 above)
Comment (ShiboXing) [reply pytorch/training/docker/1.13/py3/cu117/Dockerfile.trcomp.gpu:13]: It's a good idea to get rid of the reinstalling, including sagemaker and its training toolkit. I think Bruce maintains the structure from the last release here. We should revamp and get rid of these in PT2.0

Comment (BruceChang9688) [reply pytorch/training/docker/1.13/py3/cu117/Dockerfile.trcomp.gpu:13]: We have some new feature in sagemaker training toolkit 4.4.5 like this: https://github.com/aws/sagemaker-training-toolkit/pull/171/files. Not sure whether we need to upgrade this.

Comment (ShiboXing) [reply pytorch/training/docker/1.13/py3/cu117/Dockerfile.trcomp.gpu]: yes we will need upgrade. 
Sorry I forgot about this. I've added this upgraded in the base image revamp PR but haven't merged yet.

Comment (BruceChang9688) [reply pytorch/training/docker/1.13/py3/cu117/Dockerfile.trcomp.gpu]: Got it. Then I will remove this part in Hopper DLC.

**Review by ShiboXing — APPROVED** [review]

**Review by Lokiiiiii — APPROVED** [review]

**Comment by ShiboXing** [comment]
Could you turn off build and add trcomp skip test market to `test_pytorchddp_throughput_gpu ` and `test_smdataparallel_throughput ` for the efa? And then rerun the efa tests

**Review by ShiboXing — DISMISSED** [review]

**Comment by ShiboXing** [comment]
You can revert the developer config to be merged

**Review by ShiboXing — APPROVED** [review]

**Review by ShiboXing — APPROVED** [review]

**Review by Lokiiiiii — APPROVED** [review]

**Review by tejaschumbalkar — APPROVED** [review]

**Pull request merged**